components/cdc/src/endpoint.rs

```diff
@@ -710,29 +776,75 @@ impl<T: 'static + RaftStoreRouter<E>, E: KvEngine> Endpoint<T, E> {
+        // TODO: figure out how to avoid create binary heap and hashset
+        //       every time, saving some CPU.
+        let mut resolved_regions = ResolvedRegionHeap {
+            heap: BinaryHeap::with_capacity(regions.len()),
+        };
         self.min_resolved_ts = TimeStamp::max();
+        let mut advance_ok = 0;
+        let mut advance_failed_none = 0;
+        let mut advance_failed_same = 0;
+        let mut advance_failed_stale = 0;
         for region_id in regions {
             if let Some(delegate) = self.capture_regions.get_mut(&region_id) {
+                let old_resolved_ts = delegate
+                    .resolver
+                    .as_ref()
+                    .map_or(TimeStamp::zero(), |r| r.resolved_ts());
+                if old_resolved_ts > min_ts {
+                    advance_failed_stale += 1;
+                }
                 if let Some(resolved_ts) = delegate.on_min_ts(min_ts) {
                     if resolved_ts < self.min_resolved_ts {
                         self.min_resolved_ts = resolved_ts;
                         self.min_ts_region_id = region_id;
                     }
-                    resolved_regions.insert(region_id);
+                    resolved_regions.push(region_id, resolved_ts);
+
+                    if resolved_ts == old_resolved_ts {
+                        advance_failed_same += 1;
+                    } else {
+                        advance_ok += 1;
+                    }
+                } else {
+                    advance_failed_none += 1;
                 }
             }
         }
-        self.resolved_region_count = resolved_regions.len();
+        let lag_millis = min_ts
+            .physical()
+            .saturating_sub(self.min_resolved_ts.physical());
+        if Duration::from_millis(lag_millis) > WARN_RESOLVED_TS_LAG_THRESHOLD {
+            warn!("cdc resolved ts lag too large";
+                "min_resolved_ts" => self.min_resolved_ts,
+                "min_ts_region_id" => self.min_ts_region_id,
+                "min_ts" => min_ts,
+                "ok" => advance_ok,
+                "none" => advance_failed_none,
+                "stale" => advance_failed_stale,
+                "same" => advance_failed_same);
+        }
+        self.resolved_region_count = resolved_regions.heap.len();
         self.unresolved_region_count = total_region_count - self.resolved_region_count;
-        self.broadcast_resolved_ts(resolved_regions);
+
+        // Separate broadcasing outlier regions and normal regions,
+        // so 1) downstreams know where they should send resolve lock requests,
+        // and 2) resolved ts of normal regions does not fallback.
+        //
+        // Max number of outliers, in most cases, only a few regions are outliers.
+        let max_outlier_count = 32;
+        let (outlier_min_resolved_ts, outlier_regions) =
+            resolved_regions.pop_outliers(max_outlier_count);
```

Comment:
How about pass in a timestamp into `pop_outliers` instead of a count?

Reply:
Or maybe we could name this function `pop` rather than `pop_outliers`.